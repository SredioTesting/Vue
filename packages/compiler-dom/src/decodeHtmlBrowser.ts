/* eslint-disable no-restricted-globals */

const encodeRE = /<|>/g
const encodeMap: Record<string, string> = {
  '<': '&lt;',
  '>': '&gt;'
}

let decoder: HTMLDivElement

export function decodeHtmlBrowser(raw: string): string {
  ;(
    decoder || (decoder = document.createElement('div'))
  ).innerHTML = raw.replace(encodeRE, p1 => encodeMap[p1])
  return decoder.textContent as string
}
