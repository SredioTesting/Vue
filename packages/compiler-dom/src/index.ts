import { baseCompile, CompilerOptions, CodegenResult } from '@vue/compiler-core'
import { parserOptionsMinimal } from './parserOptionsMinimal'
import { parserOptionsStandard } from './parserOptionsStandard'
import { transformStyle } from './transforms/transformStyle'
import { transformText } from './transforms/vText'

export function compile(
  template: string,
  options: CompilerOptions = {}
): CodegenResult {
  return baseCompile(template, {
    ...options,
    ...(__BROWSER__ ? parserOptionsMinimal : parserOptionsStandard),
    nodeTransforms: [
      transformStyle,
      transformText,
      ...(options.nodeTransforms || [])
    ],
    directiveTransforms: {
      // TODO include DOM-specific directiveTransforms
      ...(options.directiveTransforms || {})
    }
  })
}

export * from '@vue/compiler-core'
