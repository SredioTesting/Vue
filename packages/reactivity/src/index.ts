export {
  ref,
  shallowRef,
  isRef,
  toRef,
  toValue,
  toRefs,
  unref,
  proxyRefs,
  customRef,
  triggerRef,
  type Ref,
  type MaybeRef,
  type MaybeRefOrGetter,
  type ToRef,
  type ToRefs,
  type UnwrapRef,
  type ShallowRef,
  type ShallowUnwrapRef,
  type RefUnwrapBailTypes,
  type CustomRefFactory,
} from './ref'
export {
  reactive,
  readonly,
  isReactive,
  isReadonly,
  isShallow,
  isProxy,
  shallowReactive,
  shallowReadonly,
  markRaw,
  toRaw,
  type Raw,
  type DeepReadonly,
  type ShallowReactive,
  type UnwrapNestedRefs,
} from './reactive'
export {
  computed,
  type ComputedRef,
  type WritableComputedRef,
  type WritableComputedOptions,
  type ComputedGetter,
  type ComputedSetter,
  type ComputedRefImpl,
} from './computed'
export {
  effect,
  stop,
  enableTracking,
  pauseTracking,
  resetTracking,
  ReactiveEffect,
  EffectFlags,
  type ReactiveEffectRunner,
  type ReactiveEffectOptions,
  type EffectScheduler,
  type DebuggerOptions,
  type DebuggerEvent,
  type DebuggerEventExtraInfo,
} from './effect'
export { trigger, track, ITERATE_KEY } from './dep'
export {
  effectScope,
  EffectScope,
  getCurrentScope,
  onScopeDispose,
} from './effectScope'
export { TrackOpTypes, TriggerOpTypes, ReactiveFlags } from './constants'
