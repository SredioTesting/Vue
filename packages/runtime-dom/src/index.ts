import {
  createRenderer,
  createHydrationRenderer,
  warn,
  RootRenderFunction,
  CreateAppFunction,
  CreateChildAppFunction,
  Renderer,
  HydrationRenderer,
  App,
  RootHydrateFunction
} from '@vue/runtime-core'
import { nodeOps } from './nodeOps'
import { patchProp, forcePatchProp } from './patchProp'
// Importing from the compiler, will be tree-shaken in prod
import { isFunction, isString, isHTMLTag, isSVGTag, extend } from '@vue/shared'

declare module '@vue/reactivity' {
  export interface RefUnwrapBailTypes {
    // Note: if updating this, also update `types/refBail.d.ts`.
    runtimeDOMBailTypes: Node | Window
  }
}

const rendererOptions = extend({ patchProp, forcePatchProp }, nodeOps)

// lazy create the renderer - this makes core renderer logic tree-shakable
// in case the user only imports reactivity utilities from Vue.
let renderer: Renderer<Element> | HydrationRenderer

let enabledHydration = false

function ensureRenderer() {
  return renderer || (renderer = createRenderer<Node, Element>(rendererOptions))
}

function ensureHydrationRenderer() {
  renderer = enabledHydration
    ? renderer
    : createHydrationRenderer(rendererOptions)
  enabledHydration = true
  return renderer as HydrationRenderer
}

// use explicit type casts here to avoid import() calls in rolled-up d.ts
export const render = ((...args) => {
  ensureRenderer().render(...args)
}) as RootRenderFunction<Element>

export const hydrate = ((...args) => {
  ensureHydrationRenderer().hydrate(...args)
}) as RootHydrateFunction

export const createApp = ((...args) => {
  const app = ensureRenderer().createApp(...args)

  if (__DEV__) {
    injectNativeTagCheck(app)
  }

  const { mount } = app
  app.mount = (containerOrSelector: Element | string): any => {
    const container = normalizeContainer(containerOrSelector)
    if (!container) return
    const component = app._component
    if (!isFunction(component) && !component.render && !component.template) {
      component.template = container.innerHTML
    }
    // clear content before mounting
    container.innerHTML = ''
    const proxy = mount(container)
    container.removeAttribute('v-cloak')
    container.setAttribute('data-v-app', '')
    return proxy
  }

  return app
}) as CreateAppFunction<Element>

export const createChildApp = ((parentApp, ...args) => {
  if (!parentApp) {
    if (__DEV__) {
      warn(`Failed to create child app: argument parentApp must be provided.`)
    }
    return
  }
  const subapp = ensureRenderer().createApp(...args)
  // methods is inherited from the father, except `mount` and `unmount` methods
  const newSubapp = extend(subapp, {
    use: parentApp.use,
    mixin: parentApp.mixin,
    component: parentApp.component,
    directive: parentApp.directive,
    provide: parentApp.provide,
    config: parentApp.config,
    _context: extend({}, parentApp._context, { app: subapp })
  }) as App<Element>

  if (__DEV__) {
    injectNativeTagCheck(newSubapp)
  }

  const { mount } = newSubapp
  newSubapp.mount = (containerOrSelector: Element | string): any => {
    const container = normalizeContainer(containerOrSelector)
    if (!container) return
    const component = newSubapp._component
    if (!isFunction(component) && !component.render && !component.template) {
      component.template = container.innerHTML
    }
    // clear content before mounting
    container.innerHTML = ''
    const proxy = mount(container, false, newSubapp._context)
    container.removeAttribute('v-cloak')
    container.setAttribute('data-v-app', '')
    return proxy
  }

  return newSubapp
}) as CreateChildAppFunction<Element>

export const createSSRApp = ((...args) => {
  const app = ensureHydrationRenderer().createApp(...args)

  if (__DEV__) {
    injectNativeTagCheck(app)
  }

  const { mount } = app
  app.mount = (containerOrSelector: Element | string): any => {
    const container = normalizeContainer(containerOrSelector)
    if (container) {
      return mount(container, true)
    }
  }

  return app
}) as CreateAppFunction<Element>

function injectNativeTagCheck(app: App) {
  // Inject `isNativeTag`
  // this is used for component name validation (dev only)
  Object.defineProperty(app.config, 'isNativeTag', {
    value: (tag: string) => isHTMLTag(tag) || isSVGTag(tag),
    writable: false
  })
}

function normalizeContainer(container: Element | string): Element | null {
  if (isString(container)) {
    const res = document.querySelector(container)
    if (__DEV__ && !res) {
      warn(`Failed to mount app: mount target selector returned null.`)
    }
    return res
  }
  return container
}

// SFC CSS utilities
export { useCssModule } from './helpers/useCssModule'
export { useCssVars } from './helpers/useCssVars'

// DOM-only components
export { Transition, TransitionProps } from './components/Transition'
export {
  TransitionGroup,
  TransitionGroupProps
} from './components/TransitionGroup'

// **Internal** DOM-only runtime directive helpers
export {
  vModelText,
  vModelCheckbox,
  vModelRadio,
  vModelSelect,
  vModelDynamic
} from './directives/vModel'
export { withModifiers, withKeys } from './directives/vOn'
export { vShow } from './directives/vShow'

// re-export everything from core
// h, Component, reactivity API, nextTick, flags & types
export * from '@vue/runtime-core'
