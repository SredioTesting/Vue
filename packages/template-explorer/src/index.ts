import * as m from 'monaco-editor'
import { watch } from '@vue/runtime-dom'
import * as compiler from './compiler'
import { initOptions } from './options'
import * as codeEditor from './codeEditor'
import * as outputEditor from './outputEditor'

declare global {
  interface Window {
    monaco: typeof m
    _deps: any
    init: () => void
  }
}

window.init = () => {
  codeEditor.initEditor()
  outputEditor.initEditor()

  // handle resize
  window.addEventListener('resize', () => {
    codeEditor.getEditor().layout()
    outputEditor.getEditor().layout()
  })

  initOptions()
  watch(compiler.reCompile)
}
