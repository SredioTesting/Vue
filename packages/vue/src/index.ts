// This entry is the "full-build" that includes both the runtime
// and the compiler, and supports on-the-fly compilation of the template option.
import { initDev } from './dev'
import { compile, CompilerOptions, CompilerError } from '@vue/compiler-dom'
import { registerRuntimeCompiler, RenderFunction, warn } from '@vue/runtime-dom'
import * as runtimeDom from '@vue/runtime-dom'
import { isString, NOOP, generateCodeFrame, extend } from '@vue/shared'
import { InternalRenderFunction } from 'packages/runtime-core/src/component'

if (__DEV__) {
  initDev()
}

const compileCache: Record<string, RenderFunction> = Object.create(null)

function compileToFunction(
  template: string | HTMLElement,
  options?: CompilerOptions
): RenderFunction {
  if (!isString(template)) {
    if (template.nodeType) {
      template = template.innerHTML
    } else {
      __DEV__ && warn(`invalid template option: `, template)
      return NOOP
    }
  }

  const key = template
  const cached = compileCache[key]
  if (cached) {
    return cached
  }

  if (template[0] === '#') {
    const el = document.querySelector(template)
    if (__DEV__ && !el) {
      warn(`Template element not found or is empty: ${template}`)
    }
    // __UNSAFE__
    // Reason: potential execution of JS expressions in in-DOM template.
    // The user must make sure the in-DOM template is trusted. If it's rendered
    // by the server, the template should not contain any user data.
    template = el ? el.innerHTML : ``
  }

  const { code } = compile(
    template,
    extend(
      {
        hoistStatic: true,
        onError: __DEV__ ? onError : undefined,
        onWarn: __DEV__ ? e => onError(e, true) : NOOP
      } as CompilerOptions,
      options
    )
  )

  function onError(err: CompilerError, asWarning = false) {
    const message = asWarning
      ? err.message
      : `Template compilation error: ${err.message}`
    const codeFrame =
      err.loc &&
      generateCodeFrame(
        template as string,
        err.loc.start.offset,
        err.loc.end.offset
      )
    warn(codeFrame ? `${message}\n${codeFrame}` : message)
  }

  // The wildcard import results in a huge object with every export
  // with keys that cannot be mangled, and can be quite heavy size-wise.
  // In the global build we know `Vue` is available globally so we can avoid
  // the wildcard object.
  // __GLOBAL__为true的情况就说明vue在全局中存在，那么就可以通过执行new Function中入参的code，运行code（规定入参为字符串，但在函数内部会自动转成js代码） // 生成了vnode
  const render = (
    __GLOBAL__ ? new Function(code)() : new Function('Vue', code)(runtimeDom) // runtimeDom为了设置全局的环境
  ) as RenderFunction

  // mark the function as runtime compiled
  ;(render as InternalRenderFunction)._rc = true

  return (compileCache[key] = render)
}

registerRuntimeCompiler(compileToFunction)

export { compileToFunction as compile }
export * from '@vue/runtime-dom'
