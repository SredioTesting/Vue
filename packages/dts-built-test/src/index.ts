import { type PropType, defineComponent } from 'vue'

const _CustomPropsNotErased = defineComponent({
  props: {},
  setup() {},
})

export const RegularComponent = defineComponent({
  props: {},
  setup() {
    return () => {}
  },
})

type MyInterface =
  | {
      a: string
    }
  | { b: string }
export const RegularComponentProps = defineComponent({
  props: {
    a: Object as () => MyInterface,
    b: Object as PropType<MyInterface>,
  },
  setup(props) {
    return () => {}
  },
})
// #8376
export const CustomPropsNotErased =
  _CustomPropsNotErased as typeof _CustomPropsNotErased & {
    foo: string
  }
