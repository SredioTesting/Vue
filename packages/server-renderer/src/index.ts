import { initDirectivesForSSR } from 'vue'
initDirectivesForSSR()

// public
export type { SSRContext } from './render'
export { renderToString } from './renderToString'
export {
  renderToSimpleStream,
  renderToNodeStream,
  pipeToNodeWritable,
  renderToWebStream,
  pipeToWebWritable,
  type SimpleReadable,
  // deprecated
  renderToStream,
} from './renderToStream'
export class MultipleErrors extends Error {
  constructor(
    message: string,
    public errors: unknown[],
  ) {
    super(message)
  }
}

// internal runtime helpers
export * from './internal'
