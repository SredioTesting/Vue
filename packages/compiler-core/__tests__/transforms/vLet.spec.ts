import { baseCompile } from '../../src'

describe('compiler: v-let transform', () => {
  function compile(content: string) {
    return baseCompile(`<div>${content}</div>`, {
      mode: 'module',
      prefixIdentifiers: true
    }).code
  }

  test('should work', () => {
    expect(compile(`<div v-let="a=1">{{a}}</div>`)).toMatchSnapshot()
  })

  test('multiple declare', () => {
    expect(compile(
      `<div v-let="a=1,  b=2">
        {{a}} {{b}}
      </div>`
    )).toMatchSnapshot()
  })

  test('nested v-let', () => {
    expect(compile(
      `<div v-let="a=1">
        <span v-let="b=1">{{a}}{{b}}</span>
      </div>`
    )).toMatchSnapshot()
  })

  test('work with variable', () => {
    expect(compile(
      `<div v-let="a=msg">
        <span v-let="b=a">{{b}}</span>
      </div>`
    )).toMatchSnapshot()
  })

  test('complex expression', () => {
    expect(compile(
      `<div v-let="a=foo+bar">
        <span v-let="b=a+baz">{{b}}</span>
      </div>`
    )).toMatchSnapshot()
  })

})
