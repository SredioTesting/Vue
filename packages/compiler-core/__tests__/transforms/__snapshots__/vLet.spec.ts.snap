// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`compiler: v-let transform complex expression 1`] = `
"import { toDisplayString as _toDisplayString, createTextVNode as _createTextVNode, createElementVNode as _createElementVNode, openBlock as _openBlock, createElementBlock as _createElementBlock } from \\"vue\\"

export function render(_ctx, _cache) {
  return (_openBlock(), _createElementBlock(\\"div\\", null, [
    ((a=_ctx.foo+_ctx.bar) => _createElementVNode(\\"div\\", null, [
      ((b=a+_ctx.baz) => _createElementVNode(\\"span\\", null, [
        _createTextVNode(_toDisplayString(b), 1 /* TEXT */)
      ]))()
    ]))()
  ]))
}"
`;

exports[`compiler: v-let transform multiple declare 1`] = `
"import { toDisplayString as _toDisplayString, createTextVNode as _createTextVNode, createElementVNode as _createElementVNode, openBlock as _openBlock, createElementBlock as _createElementBlock } from \\"vue\\"

export function render(_ctx, _cache) {
  return (_openBlock(), _createElementBlock(\\"div\\", null, [
    ((a=1,  b=2) => _createElementVNode(\\"div\\", null, [
      _createTextVNode(_toDisplayString(a) + \\" \\" + _toDisplayString(b), 1 /* TEXT */)
    ]))()
  ]))
}"
`;

exports[`compiler: v-let transform nested v-let 1`] = `
"import { toDisplayString as _toDisplayString, createTextVNode as _createTextVNode, createElementVNode as _createElementVNode, openBlock as _openBlock, createElementBlock as _createElementBlock } from \\"vue\\"

export function render(_ctx, _cache) {
  return (_openBlock(), _createElementBlock(\\"div\\", null, [
    ((a=1) => _createElementVNode(\\"div\\", null, [
      ((b=1) => _createElementVNode(\\"span\\", null, [
        _createTextVNode(_toDisplayString(a) + _toDisplayString(b), 1 /* TEXT */)
      ]))()
    ]))()
  ]))
}"
`;

exports[`compiler: v-let transform should work 1`] = `
"import { toDisplayString as _toDisplayString, createTextVNode as _createTextVNode, createElementVNode as _createElementVNode, openBlock as _openBlock, createElementBlock as _createElementBlock } from \\"vue\\"

export function render(_ctx, _cache) {
  return (_openBlock(), _createElementBlock(\\"div\\", null, [
    ((a=1) => _createElementVNode(\\"div\\", null, [
      _createTextVNode(_toDisplayString(a), 1 /* TEXT */)
    ]))()
  ]))
}"
`;

exports[`compiler: v-let transform work with variable 1`] = `
"import { toDisplayString as _toDisplayString, createTextVNode as _createTextVNode, createElementVNode as _createElementVNode, openBlock as _openBlock, createElementBlock as _createElementBlock } from \\"vue\\"

export function render(_ctx, _cache) {
  return (_openBlock(), _createElementBlock(\\"div\\", null, [
    ((a=_ctx.msg) => _createElementVNode(\\"div\\", null, [
      ((b=a) => _createElementVNode(\\"span\\", null, [
        _createTextVNode(_toDisplayString(b), 1 /* TEXT */)
      ]))()
    ]))()
  ]))
}"
`;
